[docs] rewrite load docs (#7174)
sveltejs/kit: documentation/docs/05-load.md

	export let data;

+	// we can access `data.posts` because it's returned from
+	// the parent layout `load` function
+	$: index = data.posts.findIndex(post => post.slug === $page.params.slug);
+	$: next = data.posts[index - 1];
</script>

<h1>{data.post.title}</h1>
<div>{@html data.post.content}</div>

+{#if next}
+	<p>Next post: <a href="/blog/{next.slug}">{next.title}</a></p>
+{/if}
```

> If multiple `load` functions return data with the same key, the last one 'wins'.

### $page.data

The `+page.svelte` component, and each `+layout.svelte` component above it, has access to its own data plus all the data from its parents.

In some cases, we might need the opposite — a parent layout might need to access page data or data from a child layout. For example, the root layout might want to access a `title` property returned from a `load` function in `+page.js` or `+page.server.js`. This can be done with `$page.data`:

```svelte
/// file: src/routes/+layout.svelte
<script>
	import { page } from '$app/stores';
</script>

<svelte:head>
	<title>{$page.data.title}</title>
</svelte:head>
```

Type information for `$page.data` is provided by `App.PageData`.

### Shared vs server

> As we've seen, there are two types of `load` function:

Author: ```suggestion
As we've seen, there are two types of `load` functions:
```

Reviewer: `function` singular is correct!

Reviewer 2: I've seen PRs before for the svelte.dev docs that wanted to make this sort of grammar change. I think the singular might just be intentional and be Rich's style. It also feels a little bit more formal/British to me.

Reviewer 2: Whoops, thanks for not letting me know Rich had already commented, GitHub.

#### Output

A shared `load` function can return an object containing any values, including things like custom classes and component constructors.

A server-only `load` function must return data that can be serialized with [devalue](https://github.com/rich-harris/devalue) — anything that can be represented as JSON plus things like `BigInt`, `Date`, `Map`, `Set` and `RegExp`, or repeated/cyclical references — so that it can be transported over the network.

#### When to use which

Server-only `load` functions are convenient when you need to access data directly from a database or filesystem, or need to use private environment variables.

Shared `load` functions are useful when you need to `fetch` data from an external API and don't need private credentials, since SvelteKit can get the data directly from the API rather than going via your server. They are also useful when you need to return something that can't be serialized, such as a Svelte component constructor.

> In rare cases, you might need to use both together — for example, you might need to return an instance of a custom class that was initialised with data from your server.

Author: Wondering if this would be a more real-world use case that may also help people discover how to do this:
```suggestion
In rare cases, you might need to use both together — for example, you might need to forward an authentication token from the server for subsequent use in shared `load` functions.
```

Reviewer: I don't think we'd want to encourage that pattern. If you have to go the server for an auth token _anyway_ then it makes sense to use it on the server as well — you could get all the data you need in a single hop rather than going to the server for the token then making fetches to the external API from the browser.

The thinking here was that it directly follows from the previous two paragraphs:

1. use server-only load to get stuff from your db
2. use shared load to return funky stuff
3. use both together to return funky stuff initialised with data from the db

Author: Let's keep it as-is then

- it can make relative requests on the server (ordinarily, `fetch` requires a URL with an origin when used in a server context)
- internal requests (e.g. for `+server.js` routes) go direct to the handler function when running on the server, without the overhead of an HTTP call
- during server-side rendering, the response will be captured and inlined into the rendered HTML. Note that headers will _not_ be serialized, unless explicitly included via [`filterSerializedResponseHeaders`](/docs/hooks#server-hooks-handle). Then, during hydration, the response will be read from the HTML, guaranteeing consistency and preventing an additional network request - if you got a warning in your browser console when using the browser `fetch` instead of the `load` `fetch`, this is why.

```js
/// file: src/routes/items/[id]/+page.js
/** @type {import('./$types').PageLoad} */
export async function load({ fetch, params }) {
	const res = await fetch(`/api/items/${params.id}`);
	const item = await res.json();

	return { item };
}
```

> Cookies will only be passed through if the target host is the same as the SvelteKit application or a more specific subdomain of it.

> ### Cookies and headers

Reviewer: Basically that cookies _aren't_ just a special case of headers — `cookies.set(...)` causes a `set-cookie` header to be appended, but it also updates the cookie store that accompanies the `RequestEvent`. Meanwhile `cookies.get(...)` is nothing to do with `response.headers`, it reads from `request.headers` and the event's cookie store. Headers are the _mechanism_ for dealing with cookies, but I think it's helpful to treat them as somewhat conceptually different — most headers just affect the response, but `set-cookie` affects the user agent.

In general I think it's good to be able to refer back to things that were previously covered, so that you're layering knowledge. Otherwise someone reads about `setHeader`, thinks 'great, I can use this to set a cookie', _then_ learns that actually no, they have to learn this whole other thing.

	export let data;
</script>

{#each data.posts as post}
	...
{/each}
```

You can do the same for layouts:

```js
/// file: src/routes/blog/+layout.server.js
// @filename: ambient.d.ts
declare const db: { getBlogPostsCount: () => Promise<any[]> }

// @filename: index.js
// ---cut---
/** @type {import('./$types').LayoutServerLoad} */
export async function load() {
	return {
		postsCount: await db.getBlogPostsCount()
	};
}
```

```svelte
/// file: src/routes/blog/+layout.svelte
<script>
	/** @type {import('./$types').LayoutData} */
	export let data;
</script>

<span>{data.postsCount} blog posts available</span>
<slot />
```

Top-level promises will be awaited, which makes it easy to return multiple promises without creating a waterfall:

```js
/// file: src/routes/+page.server.js
/** @type {import('./$types').PageServerLoad} */
export function load() {
	return {
		a: Promise.resolve('a'),
		b: Promise.resolve('b'),
		c: {
			value: Promise.resolve('c')
		}
	};
}
```

```svelte
<script>
	/** @type {import('./$types').PageData} */
	export let data;

	console.log(data.a); // 'a'
	console.log(data.b); // 'b'
	console.log(data.c.value); // `Promise {...}`
</script>
```

> The `load` functions related to a certain page run in parallel by default, avoiding a waterfall of requests. Once all `load` functions have returned, rendering starts. This allows layouts and pages to access data from upper layouts. In the above example, the `data` prop in `blog/+page.svelte` _also_ has access to the `postsCount` returned from `blog/+layout.server.js`. Additionally, you can access the merged data of _all_ `load` functions through `$page.data`.

Reviewer 3: Oh, I didn't realize you could get layout data from the `data` prop directly. Does this include all nested layouts as well for a page?

Author: Yes, everything above the layout/page is merged into it. Think if it like `$page.data` which stops at the layout/page

Reviewer 3: I think this could use more clarification in the docs, then, since this was a surprise to me! I thought page.data was the only way to get all merged data.

> In some environments this is derived from request headers during server-side rendering. If you're using [adapter-node](/docs/adapters#supported-environments-node-js), for example, you may need to configure the adapter in order for the URL to be correct.

### Which load function to use

The `load` function is available in both `+page.js`/`+layout.js` and `+page.server.js`/`+layout.server.js` files. They share the same name because they do the same - providing data to a layout or page. Their capabalities differ in some ways though, and it makes sense to use one or the other depending on the situation - sometimes even both.

> If the `load` function is defined in `+page.server.js` or `+layout.server.js` ("server-only `load` function") it will only run on the server, in which case it can (for example) make database calls and access private [environment variables](/docs/modules#$env-static-private), but can only return data that can be serialized with [devalue](https://github.com/rich-harris/devalue).

Reviewer 3: This sentence is a bit lengthy. Maybe:
```suggestion
If the `load` function is defined in `+page.server.js` or `+layout.server.js` ("server-only `load` function") it will only run on the server, in which case it can (for example) make database calls and access private [environment variables](/docs/modules#$env-static-private). However, it can only return data that can be serialized with [devalue](https://github.com/rich-harris/devalue).
```

Using the shared `load` function gives you more freedom when it's run on the client as you don't have to worry about isolation of requests and can use shared variables from outside (also see [shared state](#shared-state)). It's also easier to fine-tune when the `load` function reruns and avoid network requests altogether if a `load` function doesn't need to rerun or doesn't access anything that results in a network request. At the same time you have to take the isomorphic nature of the `load` function into account. If you have things you only want to do on the client or the server, use the `browser` import from `$app/environment`.

```js
/// file: src/routes/blog/+page.js
import { browser } from '$app/environment';

// this shared variable is safe to modify on the browser, as it only affects the current user
let count = 0;

/** @type {import('./$types').PageLoad} */
export async function load() {
	// Increase count each time the load function is run, but only in the browser
	const returnedCount = count;
	if (browser) {
		count++;
	}
	return { count: returnedCount };
}
```

> Sometimes it's even benefitial to use both `load` functions. For example you could first return some data from your database, and use that to create an object that is nonserializable and therefore couldn't be returned from the server `load` function. To access the server data in the shared `load` function, use the `data` input:

Reviewer 4: ```suggestion
Sometimes it's even beneficial to use both `load` functions. For example you could first return some data from your database, and use that to create an object that is non-serializable (so therefore couldn't be returned directly from the server `load` function.) To access the server data in the shared `load` function, use the `data` input:
```

// @filename: index.js
// ---cut---
/** @type {import('./$types').PageServerLoad} */
export async function load() {
	return {
		posts: await db.getAllBlogPosts()
	};
}
```

```js
/// file: src/routes/blog/+page.js
// @filename: ambient.d.ts
declare const Post: any;

// @filename: index.js
// @errors: 2531 7006
// ---cut---
/** @type {import('./$types').PageLoad} */
export async function load({ data }) {
	return {
		posts: data.posts.map(post => new Post(post))
	};
}
```

```svelte
/// file: src/routes/blog/+page.svelte
<script>
	/** @type {import('./$types').PageData} */
	export let data;
</script>

{#each data.posts as post}
	...
{/each}
```

To summarize:
- use server-only `load` in `+page.server.js`/`+layout.server.js` when
	- you want to access your data base directly
	- you access private environment variables
	- the returned data is serializable
- use shared `load` in `+page.js`/`+layout.js` when
	- you want more control over the `load` function
	- you want to avoid a network hop when the resource is on a different server
	- you need to return non-serializable data
	- you are building an SPA served from a static file server

> - use both `load` functions in case you have a combination of requirements

Reviewer 3: For consistency, maybe:
```suggestion
- use both `load` functions when you have a combination of requirements
```

> ### Making fetch requests

Author: @Conduitry in #7113 you mentioned you wanted some more details about the subtle difference of `fetch` and when to use it - this would be the section to add it to.

> Inside server-only `load` functions you can directly query your data base. Inside shared `load` functions you can't, as they also may run on the client. Also sometimes your data base may not be on the same server as your app. For these reasons it is often necessary to make `fetch` requests.

Reviewer 3: ```suggestion
Inside server-only `load` functions you can directly query your database. Inside shared `load` functions you can't, as they also may run on the client. Also sometimes your data base may not be on the same server as your app. For these reasons it is often necessary to make `fetch` requests.
```

> Cookies will only be passed through if the target host is the same as the SvelteKit application or a more specific subdomain of it.

### Headers and cookies

A `load` function can set headers and cookies on the server.

If you need to set headers for the response, you can do so using the `setHeaders` method. This is useful if you want the page to be cached, for example:

```js
// @errors: 2322
/// file: src/routes/blog/+page.server.js
/** @type {import('./$types').PageServerLoad} */
export async function load({ fetch, setHeaders }) {
	const url = `https://cms.example.com/articles.json`;
	const response = await fetch(url);

	setHeaders({
		age: response.headers.get('age'),
		'cache-control': response.headers.get('cache-control')

> 	});

Reviewer 3: This doesn't actually work right now, see #7125 and #6477

Author: Yeah I took this snippet from the old docs and later realized that, too. Not sure how important it is to switch to another example given we want to support this eventually. cc @Rich-Harris

To redirect users, use the `redirect` helper from `@sveltejs/kit` to specify the location to which they should be redirected alongside a `3xx` status code.

```js
/// file: src/routes/user/+layout.server.js
// @filename: ambient.d.ts
declare namespace App {
	interface Locals {
		user?: {
			name: string;
		}
	}
}

// @filename: index.js
// ---cut---
import { redirect } from '@sveltejs/kit';

/** @type {import('./$types').LayoutServerLoad} */
export function load({ locals }) {
	if (!locals.user) {
		throw redirect(307, '/login');
	}
}
```

### Invalidation

SvelteKit tracks the dependencies of each `load` function to avoid re-running it unnecessarily during navigation. For example, a `load` function in a root `+layout.js` doesn't need to re-run when you navigate from one page to another unless it references `url` or a member of `params` that changed since the last navigation. `load` functions also take into account parent `load` functions, if you reference them through `await parent()` - in this case, if an upper `load` function is rerun, so is the `load` function that does `await parent()`.

> `load` functions do not only run when you navigate, they can also be triggered to rerun by `invalidate` or `invalidateAll` (which we get to in the next paragraph). `invalidate` is closely connected to `fetch` and `depends`, which are both methods passed to `load`. Making a `fetch` request automatically registers the fetched URL as a dependency of the load function. `depends` does the same in a manual way and registers the specified URL as a dependency. You can then use `invalidate` and pass one of the registered URLs (or a function, if you need to decide based on a pattern rather than the full URL) to it to rerun the `load` function:

Reviewer 4: About `depends`. It took a while this morning for me to figure out that you call `depends` in the `load` _you are depending on_, not elsewhere (like a `load` further down the tree.) Maybe this was just because it was my first time using it. I'm fine with the word `depends` now, but maybe we can point out that it means (AFIACT) "define a cache key for this `load` function."
